### ISSUE TITLE
Chebfun2 can overestimate numerical degrees

### ISSUE THREAD

**Issue by ajt60gaibb** [chebfun/chebfun#1249]
Chebfun2 can get the overestimate the numerical degree of a function.

For example,  
```
f = chebfun(@(x) 1./(1+25*x.^2)); length(f)
ans = 181
f2 = chebfun2(@(x,y) 1./(1+25*x.^2)); length(f2.rows)
ans = 257
```
That's odd, but much odder is that in the y direction, 
where there is no variation at all, we get length 17 instead of 1:
```
f2 = chebfun2(@(x,y) 1./(1+25*x.^2)); length(f2.cols)
```
This may have something to do with `epslevels` because chebfun2 relies on chebfun to reduce the length of the row and column slices.

**Comment by aaustin141** [comment]
There's something other than `epslevel` that's at work here:  these oddities arise even in the world of #1584.

**Comment by ajt60gaibb** [comment]
@bhashemi, do you have an opinion about this?

**Comment by bhashemi** [comment]
I think the desicion which was made and is responsible for these oddities is that: "Chebfun2 should not `simplify` rows and columns at the end of its construction process" which in turn was made to keep Chebfun2s more accurate for some examples. Here is what we see if we do `simplify`:
`>> f3 = simplify(chebfun2(@(x,y) 1./(1+25*x.^2))); length(f3.rows), length(f3.cols)
ans =
   185
ans =
     1
`
Note that the smallest grid used in Chebfun2 constructor is 17*17 as set in `minSamples` in `chebfunpref.mergeTechPrefs` and without `simplify` the length will remain at-least 17 in both variables.

**Comment by ajt60gaibb** [comment]
Maybe Chebfun2 should simplify somehow at the end of construction?   

I cannot remember the examples that are more accurate without simplify, but I do remember seeing that.  I am not sure if it is still true though.

**Comment by trefethen** [comment]
My view in general is in favor of simplification.  Chebfun (1,2,3,...) is based on a simple story: we construct a function and then round it to 16-digits.  In my view that story should capture the truth except when there are compelling reasons to make an exception.

**Comment by ajt60gaibb** [comment]
I agree. I do not believe there is a compelling reason for an exception here. Though, @bhashemi has more experience with the improved version of the Chebfun2 constructor than me. 

I cannot remember the decision: "Chebfun2 should not `simplify` rows and columns at the end of its construction process"

**Comment by bhashemi** [comment]
From Chebfun2 paper published in SISC: Remark. We deliberately do not represent each column and row slice as an independent chebfun "with the minimal degree".
This is what I meant by `the desicion`. The paper mentions efficiency (sorry, not accuracy) as the reason to do so.

Anyway, I am also in favor of doing `simplify` at the end of Chebfun2 construction process.

**Comment by ajt60gaibb** [comment]
This is very different.  In that remark in the paper we mean that the all the columns (rows) are simplified to the same length. The next sentence in the remark we clarify this: 
```
We deliberately do not represent each column and row slice as an independent chebfun with the minimal 
degree. Instead we ensure all the column slices have the same degree. Similarly, all the row slices 
have the same degree (which can be different than the degree of the column slices). 
```

Here we are discussing why the columns (rows) seem to not be simplified at all (!)

**Comment by bhashemi** [comment]
@ajt60gaibb, here is Line 380 (Phase II) of `@chebfun2/constructor` 
`resolvedRows = happinessCheck(rowChebtech,[],sum(rowValues.',2));`
Now consider your first example above:
`f2 = chebfun2(@(x,y) 1./(1+25*x.^2));`
We have `resolvedRows = 1` once `numel(sum(rowValues.',2)) = 257`. We don't care what the cutoff output of happinessCheck is and there is no further chopping. But, if we say
`[resolvedRows,~,cutoff] = happinessCheck(rowChebtech,[],sum(rowValues.',2))`
we get
`resolvedRows =
     1`
`cutoff =
   189`
which shows where chebfun2 can chop. We can fix this right here but (since a similar story holds for resolvedCols and also in Phase 1) a better place might be a call to `simplify` at the very end of the construction. So, here is one possibility: Replace Lines  406-407 of `@chebfun2/constructor` which are
`g.cols = chebfun(colValues, dom(3:4), pref);`
`g.rows = chebfun(rowValues.', dom(1:2), pref );`
with
`g.cols = simplify(chebfun(colValues, dom(3:4), pref));`
`g.rows = simplify(chebfun(rowValues.', dom(1:2), pref ));`
This fixes the above issues as we now get
`f2 = chebfun2(@(x,y) 1./(1+25*x.^2)); length(f2.rows), length(f2.cols)`
`ans =
   185`
`ans =
     1`

This is what I used to do in `Chebfun3sTu`:
`>> f3 = chebfun3sTu(@(x,y,z) 1./(1+25*x.^2)); length(f3)`
`ans =
   185     1     1`
and in general seems to work well for other examples also.

**Comment by ajt60gaibb** [comment]
Yes, I think `simplify` should be called at the end of the constructor like you do in `Chebfun3sTu`.  That seems great.

**Comment by aaustin141** [comment]
Closed with #1644.
